File: manager/procd/pkg/http/handlers/initialize.go
package handlers

import (
	"encoding/json"
	"net/http"
	"strings"
	"sync"

	ctxpkg "github.com/sandbox0-ai/sandbox0/manager/procd/pkg/context"
	"github.com/sandbox0-ai/sandbox0/manager/procd/pkg/file"
	"github.com/sandbox0-ai/sandbox0/manager/procd/pkg/webhook"
	"github.com/sandbox0-ai/sandbox0/pkg/internalauth"
	"go.uber.org/zap"
)

// InitializeHandler handles sandbox initialization requests.
type InitializeHandler struct {
	dispatcher     *webhook.Dispatcher
	fileManager    *file.Manager
	contextManager *ctxpkg.Manager
	httpPort       int
	logger         *zap.Logger
	readyOnce      sync.Once
	watchMu        sync.Mutex
	watchPath      string
	unsubscribe    func() error
}

// NewInitializeHandler creates a new initialize handler.
func NewInitializeHandler(dispatcher *webhook.Dispatcher, fileManager *file.Manager, contextManager *ctxpkg.Manager, httpPort int, logger *zap.Logger) *InitializeHandler {
	return &InitializeHandler{
		dispatcher:     dispatcher,
		fileManager:    fileManager,
		contextManager: contextManager,
		httpPort:       httpPort,
		logger:         logger,
	}
}

// InitializeRequest is the request body for initializing sandbox settings.
type InitializeRequest struct {
	SandboxID string             `json:"sandbox_id"`
	TeamID    string             `json:"team_id,omitempty"`
	EnvVars   map[string]string  `json:"env_vars,omitempty"`
	Webhook   *InitializeWebhook `json:"webhook,omitempty"`
}

// InitializeWebhook represents webhook configuration.
type InitializeWebhook struct {
	URL      string `json:"url"`
	Secret   string `json:"secret,omitempty"`
	WatchDir string `json:"watch_dir,omitempty"`
}

// InitializeResponse is returned after initialization.
type InitializeResponse struct {
	SandboxID string `json:"sandbox_id"`
	TeamID    string `json:"team_id,omitempty"`
}

// UpdateSandboxEnvVarsRequest updates sandbox-level default environment variables.
type UpdateSandboxEnvVarsRequest struct {
	EnvVars map[string]string `json:"env_vars,omitempty"`
}

// UpdateSandboxEnvVarsResponse returns the normalized sandbox environment.
type UpdateSandboxEnvVarsResponse struct {
	EnvVars map[string]string `json:"env_vars"`
}

// Initialize sets sandbox identity and webhook settings.
func (h *InitializeHandler) Initialize(w http.ResponseWriter, r *http.Request) {
	if h.dispatcher == nil {
		writeError(w, http.StatusServiceUnavailable, "webhook_unavailable", "webhook dispatcher not configured")
		return
	}

	var req InitializeRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid_request", err.Error())
		return
	}

	if req.SandboxID == "" {
		writeError(w, http.StatusBadRequest, "invalid_request", "sandbox_id is required")
		return
	}

	claims := internalauth.ClaimsFromContext(r.Context())
	teamID := req.TeamID
	if claims != nil {
		if teamID == "" {
			teamID = claims.TeamID
		} else if claims.TeamID != "" && teamID != claims.TeamID {
			writeError(w, http.StatusForbidden, "forbidden", "team_id does not match token")
			return
		}
	}

	webhookURL := ""
	webhookSecret := ""
	webhookWatchDir := ""
	if req.Webhook != nil {
		webhookURL = req.Webhook.URL
		webhookSecret = req.Webhook.Secret
		webhookWatchDir = req.Webhook.WatchDir
	}

	if h.contextManager != nil {
		h.contextManager.SetSandboxEnvVars(req.EnvVars)
	}

	h.dispatcher.SetConfig(webhookURL, webhookSecret)
	h.dispatcher.SetIdentity(req.SandboxID, teamID)

	h.configureWebhookWatch(strings.TrimSpace(webhookURL), strings.TrimSpace(webhookWatchDir))

	if webhookURL != "" {
		h.readyOnce.Do(func() {
			if _, err := h.dispatcher.Enqueue(webhook.Event{
				EventType: webhook.EventTypeSandboxReady,
				Payload: map[string]any{
					"http_port":  h.httpPort,
					"sandbox_id": req.SandboxID,
				},
			}); err != nil && h.logger != nil {
				h.logger.Warn("Failed to enqueue sandbox ready webhook", zap.Error(err))
			}
		})
	}

	writeJSON(w, http.StatusOK, InitializeResponse{
		SandboxID: req.SandboxID,
		TeamID:    teamID,
	})
}

// UpdateSandboxEnvVars updates default environment variables for future sandbox processes.
func (h *InitializeHandler) UpdateSandboxEnvVars(w http.ResponseWriter, r *http.Request) {
	var req UpdateSandboxEnvVarsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid_request", err.Error())
		return
	}
	envVars := map[string]string{}
	if h.contextManager != nil {
		h.contextManager.SetSandboxEnvVars(req.EnvVars)
		if current := h.contextManager.SandboxEnvVars(); current != nil {
			envVars = current
		}
	}
	writeJSON(w, http.StatusOK, UpdateSandboxEnvVarsResponse{EnvVars: envVars})
}

func (h *InitializeHandler) configureWebhookWatch(webhookURL, watchDir string) {
	if h.fileManager == nil {
		return
	}

	h.watchMu.Lock()
	defer h.watchMu.Unlock()

	if webhookURL == "" || watchDir == "" {
		if h.unsubscribe != nil {
			_ = h.unsubscribe()
		}
		h.unsubscribe = nil
		h.watchPath = ""
		return
	}

	if watchDir == h.watchPath {
		return
	}

	if h.unsubscribe != nil {
		_ = h.unsubscribe()
		h.unsubscribe = nil
		h.watchPath = ""
	}

	_, unsubscribe, err := h.fileManager.SubscribeWatch(watchDir, true, func(event file.WatchEvent) {
		if event.Type == file.EventInvalidate {
			return
		}
		payload := map[string]any{
			"event_type": event.Type,
			"path":       event.Path,
		}
		if event.OldPath != "" {
			payload["old_path"] = event.OldPath
		}
		if _, err := h.dispatcher.Enqueue(webhook.Event{
			EventType: webhook.EventTypeFileModified,
			Payload:   payload,
		}); err != nil && h.logger != nil {
			h.logger.Warn("Failed to enqueue file webhook", zap.Error(err))
		}
	})
	if err != nil {
		if h.logger != nil {
			h.logger.Warn("Failed to watch webhook directory",
				zap.String("watch_dir", watchDir),
				zap.Error(err),
			)
		}
		return
	}

	h.unsubscribe = unsubscribe
	h.watchPath = watchDir
}

File: manager/procd/src/http/handlers/initialize.js
import fs from "node:fs";
import path from "node:path";
import { readJSON, writeError, writeJSON } from "../response.js";
import { eventTypes } from "../../webhook/dispatcher.js";

function ensureMountDirs(dirs) {
  for (let i = 0; i < (dirs || []).length; i++) {
    const dir = path.resolve(String(dirs[i] || "").trim());
    if (!path.isAbsolute(dir) || dir === path.parse(dir).root) {
      throw new Error(`mount_dirs[${i}] must be an absolute non-root path`);
    }
    fs.mkdirSync(dir, { recursive: true, mode: 0o755 });
  }
}

export class InitializeHandler {
  constructor(webhookDispatcher, fileManager, contextManager, httpPort, logger) {
    this.webhookDispatcher = webhookDispatcher;
    this.fileManager = fileManager;
    this.contextManager = contextManager;
    this.httpPort = httpPort;
    this.logger = logger;
    this.readySentKey = "";
    this.unsubscribe = null;
    this.watchPath = "";
  }

  async initialize(req, res) {
    let body;
    try {
      body = await readJSON(req);
      if (!body.sandbox_id) {
        writeError(res, 400, "invalid_request", "sandbox_id is required");
        return;
      }
      ensureMountDirs(body.mount_dirs || []);
    } catch (err) {
      writeError(res, err.message?.startsWith("mount_dirs") ? 500 : 400, err.message?.startsWith("mount_dirs") ? "mount_dir_init_failed" : "invalid_request", err.message);
      return;
    }
    const claims = req.internalClaims || {};
    let teamID = body.team_id || "";
    if (claims.team_id) {
      if (!teamID) teamID = claims.team_id;
      else if (teamID !== claims.team_id) {
        writeError(res, 403, "forbidden", "team_id does not match token");
        return;
      }
    }
    this.contextManager.setSandboxEnvVars(body.env_vars || {});
    const webhook = body.webhook || {};
    this.webhookDispatcher.setConfig(webhook.url || "", webhook.secret || "");
    this.webhookDispatcher.setIdentity(body.sandbox_id, teamID);
    this.configureWebhookWatch((webhook.url || "").trim(), (webhook.watch_dir || "").trim());
    if (webhook.url) {
      const readyKey = `${body.sandbox_id}\0${webhook.url}`;
      if (readyKey !== this.readySentKey) {
        try {
          this.webhookDispatcher.enqueue({
            event_type: eventTypes.sandboxReady,
            payload: { http_port: this.httpPort, sandbox_id: body.sandbox_id }
          });
          this.readySentKey = readyKey;
        } catch (err) {
          writeError(res, 503, "webhook_enqueue_failed", err.message);
          return;
        }
      }
    }
    writeJSON(res, 200, { sandbox_id: body.sandbox_id, team_id: teamID });
  }

  async updateSandboxEnvVars(req, res) {
    try {
      const body = await readJSON(req);
      this.contextManager.setSandboxEnvVars(body.env_vars || {});
      writeJSON(res, 200, { env_vars: this.contextManager.getSandboxEnvVars() });
    } catch (err) {
      writeError(res, 400, "invalid_request", err.message);
    }
  }

  configureWebhookWatch(webhookURL, watchDir) {
    if (!webhookURL || !watchDir) {
      if (this.unsubscribe) this.unsubscribe();
      this.unsubscribe = null;
      this.watchPath = "";
      return;
    }
    if (watchDir === this.watchPath) return;
    if (this.unsubscribe) this.unsubscribe();
    this.unsubscribe = null;
    this.watchPath = "";
    try {
      const { unsubscribe } = this.fileManager.subscribeWatch(watchDir, true, (event) => {
        if (event.type === "invalidate") return;
        const payload = { event_type: event.type, path: event.path };
        if (event.old_path) payload.old_path = event.old_path;
        this.webhookDispatcher.enqueue({ event_type: eventTypes.fileModified, payload });
      });
      this.unsubscribe = unsubscribe;
      this.watchPath = watchDir;
    } catch (err) {
      this.logger?.warn("Failed to watch webhook directory", { watch_dir: watchDir, error: err.message });
    }
  }
}
